internal/app/controller/handlers.go

```diff
@@ -0,0 +1,404 @@
+	"go-developer-course-diploma/internal/app/storage/repository"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+)
+
+const (
+	NEW        = "NEW"
+	PROCESSING = "PROCESSING"
+	INVALID    = "INVALID"
+	PROCESSED  = "PROCESSED"
+)
+
+func WriteError(w http.ResponseWriter, code int, err error) {
+	WriteResponse(w, code, err.Error())
+}
+
+func WriteResponse(w http.ResponseWriter, statusCode int, data string) {
+	w.WriteHeader(statusCode)
+	if len(data) != 0 {
+		w.Write([]byte(data))
+	}
+}
+
+func WriteJSON(w http.ResponseWriter, response interface{}, logger *logrus.Logger) {
+	buf := bytes.NewBuffer([]byte{})
+	encoder := json.NewEncoder(buf)
+	err := encoder.Encode(response)
+	if err != nil {
+		logger.Infof("Encoder error: %s", err)
+		WriteError(w, http.StatusInternalServerError, err)
+		return
+	}
+
+	logger.Debugf("WriteJSON response: %s", buf.String())
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+
+	_, err = w.Write(buf.Bytes())
+	if err != nil {
+		WriteError(w, http.StatusInternalServerError, err)
+		return
+	}
+}
+
+func IsValidOrderNumber(number string) bool {
+	value, err := strconv.Atoi(number)
+	if err != nil {
+		return false
+	}
+	return luhn.Valid(value)
+}
+
+func getPasswordHash(u *model.User) {
+	u.Password = hex.EncodeToString([]byte(u.Password))
+}
+
+type Controller struct {
+	Config                 *configs.Config
+	Logger                 *logrus.Logger
+	UserRepository         repository.UserRepository
+	OrderRepository        repository.OrderRepository
+	TransactionRepository  repository.TransactionRepository
+	UserAuthorizationStore secure.UserAuthorization
+}
+
+func NewController(cfg *configs.Config, logger *logrus.Logger, userStore repository.UserRepository, orderStore repository.OrderRepository, transactionStore repository.TransactionRepository, userAuthorizationStore secure.UserAuthorization) *Controller {
+	return &Controller{
+		Config:                 cfg,
+		Logger:                 logger,
+		UserRepository:         userStore,
+		OrderRepository:        orderStore,
+		TransactionRepository:  transactionStore,
+		UserAuthorizationStore: userAuthorizationStore,
+	}
+}
+
+func (c *Controller) RegisterHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var user *model.User
+		if err := json.NewDecoder(r.Body).Decode(&user); err != nil {
+			WriteError(w, http.StatusBadRequest, err)
+			return
+		}
+
+		if len(user.Login) == 0 || len(user.Password) == 0 {
+			WriteResponse(w, http.StatusBadRequest, "")
+			return
+		}
+
+		getPasswordHash(user)
+
+		err := c.UserRepository.RegisterUser(user)
+		if err != nil && !errors.Is(err, repository.ErrorUserAlreadyExist) {
+			c.Logger.Infof("RegisterUser error: %s", err)
+			WriteError(w, http.StatusInternalServerError, err)
+			return
+		}
+		if errors.Is(err, repository.ErrorUserAlreadyExist) {
+			WriteError(w, http.StatusConflict, err)
+			return
+		}
```

Comment:
Можно переставить порядок, тогда кучу условий с errors.Is не надо `errors.Is(nil, repository.ErrorUserAlreadyExist)` всегда будет false :)
```suggestion
		if errors.Is(err, repository.ErrorUserAlreadyExist) {
			WriteError(w, http.StatusConflict, err)
			return
		}
		if err != nil  {
			c.Logger.Infof("RegisterUser error: %s", err)
			WriteError(w, http.StatusInternalServerError, err)
			return
		}
```

Reply:
Исправил здесь и в остальных похожих местах :)